Index: ./indra/llwindow/llwindowsdl.cpp
===================================================================
--- ./indra/llwindow/llwindowsdl.cpp	(revision 57)
+++ ./indra/llwindow/llwindowsdl.cpp	(working copy)
@@ -135,7 +135,7 @@
 			<< gtk_major_version << "."
 			<< gtk_minor_version << "."
 			<< gtk_micro_version << llendl;
-		gchar *gtk_warning;
+		const gchar *gtk_warning;
 		maybe_lock_display();
 		gtk_warning = gtk_check_version(GTK_MAJOR_VERSION,
 						GTK_MINOR_VERSION,
