--- llcurl.h.anc	2008-03-04 18:24:32.000000000 +0100
+++ newview/indra/llmessage/llcurl.h	2008-03-04 18:24:52.000000000 +0100
@@ -150,8 +150,8 @@
 	static std::vector<LLMutex*> sSSLMutex;
 
 	// OpenSSL callbacks
-	static void LLCurl::ssl_locking_callback(int mode, int type, const char *file, int line);
-	static unsigned long LLCurl::ssl_thread_id(void);
+	static void ssl_locking_callback(int mode, int type, const char *file, int line);
+	static unsigned long ssl_thread_id(void);
 	
 	
 	
