--- llpanelweb.cpp.anc	2008-01-24 21:03:21.000000000 +0100
+++ linden/indra/newview/llpanelweb.cpp	2008-01-24 21:05:15.000000000 +0100
@@ -43,7 +43,6 @@
 #include "llbutton.h"
 #include "llcheckboxctrl.h"
 #include "lllineeditor.h"
-#include "llmozlib.h"
 #include "llui.h"
 #include "lluictrlfactory.h"
 #include "llviewercontrol.h"
@@ -124,8 +123,9 @@
 	gSavedSettings.setBOOL( "BrowserProxyEnabled", mWebProxyEnabled );
 	gSavedSettings.setString( "BrowserProxyAddress", mWebProxyAddress );
 	gSavedSettings.setS32( "BrowserProxyPort", mWebProxyPort );
-
+#if LL_LIBXUL_ENABLED
 	LLMozLib::getInstance()->enableProxy( mWebProxyEnabled, mWebProxyAddress, mWebProxyPort ); 
+#endif
 }
 
 // static
@@ -195,10 +195,11 @@
 			check->get());
   self->childSetEnabled("proxy_text_label", 
 			check->get());
-	
+#if LL_LIBXUL_ENABLED
   LLMozLib::getInstance()->enableProxy( gSavedSettings.getBOOL("BrowserProxyEnabled"), 
 										  gSavedSettings.getString("BrowserProxyAddress"), 
 										  gSavedSettings.getS32("BrowserProxyPort") ); 
+#endif
 }
 
 void LLPanelWeb::onCommitWebProxyAddress(LLUICtrl *ctrl, void *userdata)
@@ -209,16 +210,19 @@
   {
 	  gSavedSettings.setString("BrowserProxyAddress", web_proxy->getText());
   }	
-  
+#if LL_LIBXUL_ENABLED
   LLMozLib::getInstance()->enableProxy( gSavedSettings.getBOOL("BrowserProxyEnabled"), 
 										  gSavedSettings.getString("BrowserProxyAddress"), 
 										  gSavedSettings.getS32("BrowserProxyPort") ); 
+#endif
 }
 
 void LLPanelWeb::onCommitWebProxyPort(LLUICtrl *ctrl, void *userdata)
 {
+#if LL_LIBXUL_ENABLED
 	LLMozLib::getInstance()->enableProxy( gSavedSettings.getBOOL("BrowserProxyEnabled"), 
 										  gSavedSettings.getString("BrowserProxyAddress"), 
 										  gSavedSettings.getS32("BrowserProxyPort") ); 
 
+#endif
 }
