--- llmimetypes.h.anc	2008-02-28 20:44:05.000000000 +0100
+++ linden/indra/newview/llmimetypes.h	2008-02-28 20:45:04.000000000 +0100
@@ -68,7 +68,7 @@
 	static bool findAllowResize(const LLString& mime_type);
 		// accessor for flag to enable/disable media size edit fields
 
-	static bool LLMIMETypes::findAllowLooping(const LLString& mime_type);
+	static bool findAllowLooping(const LLString& mime_type);
 		// accessor for flag to enable/disable media looping checkbox
 
 public:
