diff -Naur linden.orig/indra/SConstruct linden/indra/SConstruct
--- linden.orig/indra/SConstruct	2007-08-30 06:10:39.000000000 +0900
+++ linden/indra/SConstruct	2007-08-31 23:37:53.000000000 +0900
@@ -154,6 +154,9 @@
     'libssl',
     ]
 
+if enable_gstreamer:
+	standalone_pkgs += [ 'gstreamer-0.10' ]
+
 def pkgconfig(opt, pkgs=None):
     if pkgs is None:
         pkgs = standalone_pkgs + standalone_net_pkgs
@@ -610,7 +613,7 @@
         net_external_libs = [ 'curl', 'ssl', 'crypto', 'aprutil-1', 'apr-1' ]
     net_external_libs += [ 'cares', 'expat' ]
 
-    common_external_libs = net_external_libs + [ 'xmlrpc-epi', 'z' ]
+    common_external_libs = net_external_libs + [ 'libxmlrpc-epi', 'z' ]
 
     if build_target == 'client':
         if platform == 'linux':
@@ -754,7 +757,7 @@
                             internal_libs + external_libs)
 
         # Rpcserver
-        external_libs = common_external_libs + ['xmlrpc-epi', 'mysqlclient']
+        external_libs = common_external_libs + ['libxmlrpc-epi', 'mysqlclient']
         internal_libs = ['llscene', 'llmessage', 'lldatabase', 'llvfs', 
                 'llmath', 'llcommon']
         create_executable('rpcserver/rpcserver' + file_suffix, 'rpcserver',
diff -Naur linden.orig/indra/llimagej2coj/llimagej2coj.cpp linden/indra/llimagej2coj/llimagej2coj.cpp
--- linden.orig/indra/llimagej2coj/llimagej2coj.cpp	2007-08-30 06:10:39.000000000 +0900
+++ linden/indra/llimagej2coj/llimagej2coj.cpp	2007-08-31 00:33:42.000000000 +0900
@@ -31,6 +31,6 @@
 
 // this is defined so that we get static linking.
-#include "openjpeg/openjpeg.h"
+#include "openjpeg.h"
 
 #include "lltimer.h"
 #include "llmemory.h"
diff -Naur linden.orig/indra/newview/llvoiceclient.cpp linden/indra/newview/llvoiceclient.cpp
--- linden.orig/indra/newview/llvoiceclient.cpp	2007-08-30 06:10:43.000000000 +0900
+++ linden/indra/newview/llvoiceclient.cpp	2007-08-31 00:33:42.000000000 +0900
@@ -36,7 +36,7 @@
 #include "llvoavatar.h"
 #include "llbufferstream.h"
 #include "llfile.h"
-#include "expat/expat.h"
+#include "expat.h"
 #include "llcallbacklist.h"
 #include "llviewerregion.h"
 #include "llviewernetwork.h"		// for gUserServerChoice
diff -Naur linden.orig/indra/newview/llwebbrowserctrl.h linden/indra/newview/llwebbrowserctrl.h
--- linden.orig/indra/newview/llwebbrowserctrl.h	2007-08-30 06:10:43.000000000 +0900
+++ linden/indra/newview/llwebbrowserctrl.h	2007-08-31 00:33:42.000000000 +0900
@@ -92,6 +92,7 @@
 #include "lldynamictexture.h"
 #include "llmozlib.h"
 
+class LLUICtrlFactory;
 class LLViewBorder;
 class LLWebBrowserTexture;
 
