diff -urN --exclude='*~' OpenJPEG/libopenjpeg/t1.h OpenJPEG.orig/libopenjpeg/t1.h
--- OpenJPEG/libopenjpeg/t1.h	2007-01-23 05:41:16.000000000 -0600
+++ OpenJPEG.orig/libopenjpeg/t1.h	2007-04-02 01:48:46.000000000 -0500
@@ -44,8 +44,8 @@
 /* ----------------------------------------------------------------------- */
 #define T1_NMSEDEC_BITS 7
 
-#define T1_MAXCBLKW 1024	/**< Maximum size of code-block (width) */
-#define T1_MAXCBLKH 1024	/**< Maximum size of code-block (heigth) */
+#define T1_MAXCBLKW 64		/**< Maximum size of code-block (width) */
+#define T1_MAXCBLKH 64		/**< Maximum size of code-block (heigth) */
 
 #define T1_SIG_NE 0x0001	/**< Context orientation : North-East direction */
 #define T1_SIG_SE 0x0002	/**< Context orientation : South-East direction */
