--- PG.conf	2005-04-01 20:59:43.000000000 +0000
+++ PG.conf	2005-06-09 13:13:33.000000000 +0000
@@ -4,7 +4,7 @@
 #  
 
 # Modify the block file below to match the block file you use
-BLOCKFILE=/etc/p2p.p2b.p2p
+BLOCKFILE=/usr/share/peerguardian/p2p.p2b.p2p
 
 # Please edit below if you wish not to block certain ports with 
 # were in the format 80,25,x,x; or leave it blank if you don't wish

--- AUTHORS	2005-04-01 18:09:45.000000000 +0000
+++ AUTHORS	2005-06-09 15:39:49.000000000 +0000
@@ -1 +1,3 @@
+Gentoo adapted changes were made by djh-world (Daniel Harper) and iainel 
+
 peerguardian for linux was created by bitrot, nullcontext and JFM.

--- ChangeLog 2005-04-01 18:08:51.000000000 +0000
+++ ChangeLog 2005-06-09 15:39:10.000000000 +0000
@@ -1 +1,16 @@
+09/06/2005: Gentoo adapated release - Daniel Harper 2005
+Release: gentoo-1.5.1_beta 
+
+* Ebuild created. 
+* pgtext.cpp changed to accomodate new file locations
+* File structure changed, config file is located in /usr/share/peerguardian
+* Blocklists are downloaded to /usr/share/peerguardian
+* pgupdate script included to update new blocklist (courtesy of Daniel Harper)  
+* pgupdate script stops and starts the daemon 
+* Gentoo initscript (courtesy of iainel) added
+* PG.conf is edited to accomodate new location
+* Designed to be fully automated once placed on boot
+* Initscript will activate the daemon and it can simply run in the background
+
+
 initial release 1.5Beta

--- pgtext.cpp 2005-03-31 05:55:36.000000000 +0000
+++ pgtext.cpp 2005-06-09 15:13:05.000000000 +0000
@@ -32,9 +32,9 @@
 #include <limits.h>
 
 #define FIFO_NAME "/tmp/PG_fifo"
-#define DEFAULT_BLOCKLIST "/etc/p2p.p2b.p2p"
+#define DEFAULT_BLOCKLIST "/usr/share/peerguardian/p2p.p2b.p2p"
 #define DEFAULT_LOG "/var/log/PG.log"
-#define DEFAULT_CONF "/etc/PG.conf"
+#define DEFAULT_CONF "/usr/share/peerguardian/PG.conf"
 #define DEFAULT_EDITOR "vi"
 #define BUFFER_SIZE PIPE_BUF
 
@@ -46,9 +46,9 @@
 int response;
 
 // various file name variables
-char blocklist_name[25] = DEFAULT_BLOCKLIST;
+char blocklist_name[40] = DEFAULT_BLOCKLIST;
 char log_name[25] = DEFAULT_LOG;
-char conf_name[25] = DEFAULT_CONF;
+char conf_name[40] = DEFAULT_CONF;
 char editor_name[25] = DEFAULT_EDITOR;
 
 using namespace std;
