--- Mozlib.cmake.anc	2009-03-27 21:59:41.265163362 +0100
+++ indra/cmake/Mozlib.cmake	2009-03-27 22:00:30.353011814 +0100
@@ -19,13 +19,14 @@
         link_directories(/usr/lib/llmozlib2)
         set(MOZLIB_LIBRARIES
             llmozlib2
-            mozjs
-            nspr4
-            plc4
-            plds4
-            xpcom
-            xul
-            profdirserviceprovider_s
+            qgif
+            qjpeg
+            QtWebKit
+            QtOpenGL
+            QtNetwork
+            QtGui
+            QtCore
+            fontconfig
             )
     elseif (WINDOWS)
         if (MSVC71)
