--- llvoiceclient.cpp.anc	2009-05-12 18:01:06.000000000 +0200
+++ indra/newview/llvoiceclient.cpp	2009-05-12 18:01:38.000000000 +0200
@@ -1588,10 +1588,6 @@
 						parcelChanged();
 					}
 				}
-				else
-				{
-					LL_WARNS("Voice") << "region doesn't have ParcelVoiceInfoRequest capability.  This is normal for a short time after teleporting, but bad if it persists for very long." << LL_ENDL;
-				}
 			}
 		}
 	}
